redwood-treehouse-host/src/commonMain/kotlin/app/cash/redwood/treehouse/CodeHost.kt

```diff
@@ -15,16 +15,184 @@
+    state = State.Starting(codeUpdatesScope)
+    codeUpdatesScope.collectCodeUpdates()
+  }
+
+  /** This function may only be invoked on [TreehouseDispatchers.zipline]. */
+  fun stop() {
+    dispatchers.checkUi()
+
+    val previous = state
+    previous.codeUpdatesScope?.cancel()
+    previous.codeSession?.removeListener(codeSessionListener)
+    previous.codeSession?.cancel()
+
+    state = State.Idle()
+  }
+
+  fun restart() {
+    dispatchers.checkUi()
+
+    val previous = state
+    if (previous is State.Starting) return // Nothing to restart.
+
+    previous.codeUpdatesScope?.cancel()
+    previous.codeSession?.removeListener(codeSessionListener)
+    previous.codeSession?.cancel()
+
+    val codeUpdatesScope = codeUpdatesScope()
+    state = State.Starting(codeUpdatesScope)
+    codeUpdatesScope.collectCodeUpdates()
+  }
+
+  fun addListener(listener: Listener<A>) {
+    dispatchers.checkUi()
+    listeners += listener
+  }
+
+  fun removeListener(listener: Listener<A>) {
+    dispatchers.checkUi()
+    listeners -= listener
+  }
+
+  private fun codeUpdatesScope() =
+    CoroutineScope(SupervisorJob(appScope.coroutineContext.job))
+
+  private fun CoroutineScope.collectCodeUpdates() {
+    launch(dispatchers.zipline) {
+      codeUpdatesFlow().collect {
+        codeSessionLoaded(it)
+      }
+    }
+  }
+
+  private fun codeSessionLoaded(next: CodeSession<A>) {
+    dispatchers.checkZipline()
+
+    next.scope.launch(dispatchers.ui) {
+      // Clean up the previous session.
+      val previous = state
+      previous.codeSession?.removeListener(codeSessionListener)
+      previous.codeSession?.cancel()
+
+      // If the codeUpdatesScope is null, we're stopped. Discard the newly-loaded code.
+      val codeUpdatesScope = state.codeUpdatesScope
+      if (codeUpdatesScope == null) {
+        next.cancel()
+        return@launch
+      }
+
+      // Boot up the new code.
+      state = State.Running(codeUpdatesScope, next)
+      next.addListener(codeSessionListener)
+      next.start()
+
+      for (listener in listeners) {
+        listener.codeSessionChanged(next)
+      }
+    }
+  }
+
+  private sealed class State<A : AppService> {
+    /** Non-null if we're prepared for code updates and restarts. */
+    open val codeUpdatesScope: CoroutineScope?
+      get() = null
+
+    /** Non-null if we're running code. */
+    open val codeSession: CodeSession<A>?
+      get() = null
 
-  fun addListener(listener: Listener<A>)
+    class Idle<A : AppService> : State<A>()
 
-  fun removeListener(listener: Listener<A>)
+    class Running<A : AppService>(
+      override val codeUpdatesScope: CoroutineScope,
+      override val codeSession: CodeSession<A>,
+    ) : State<A>()
+
+    class Starting<A : AppService>(
+      override val codeUpdatesScope: CoroutineScope,
+    ) : State<A>()
+
+    class Crashed<A : AppService>(
+      override val codeUpdatesScope: CoroutineScope,
+    ) : State<A>()
+  }
```

Comment:
Personally, I would've loved to see the states hoisted inside the respective states, in order to get the true benefits of making a state machine (i.e., it being easy to follow how one state gets to the next, a thing which is still difficult in this PR).

Apologies if that was a planned follow up!

Reply:
Yeah lemme explore that in follow up.

Reply:
A month later . . . we looked into this, and didn’t find something I loved!

@veyndan if you’d like to run with this, please do!